<a href="https://colab.research.google.com/github/ErikSeguinte/movie_data/blob/master/processing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import cufflinks as cf
import numpy as np
import plotly.graph_objects as go
import altair as alt
import plotly.express as px

In [0]:
merge_keys = {'left_index':True, 'right_index':True}

* I previously pulled CSV files from Kaggle, but the files were too big to host on github.
* I imported the files I wanted into pandas, and then exported them back out as compressed pickles.
* I was able to compress a 700MB csv to a 3 MB Pickle

In [3]:
#try: 
#    movies = pd.read_pickle('data/movies.pkl.xz')
#    ratings = pd.read_pickle('data/ratings2.pkl.xz')
#    cpi = pd.read_csv('data/cpi.csv')
    
#except FileNotFoundError                         :
    
# Download pickles from github
cpi = pd.read_csv('https://datahub.io/core/cpi/r/cpi.csv')

!wget 'https://github.com/ErikSeguinte/movie_data/raw/master/data/movies.pkl.xz'
!wget 'https://github.com/ErikSeguinte/movie_data/raw/master/data/ratings2.pkl.xz'
movies = pd.read_pickle('movies.pkl.xz')
ratings = pd.read_pickle('ratings2.pkl.xz')

--2020-03-06 03:41:35--  https://github.com/ErikSeguinte/movie_data/raw/master/data/movies.pkl.xz
Resolving github.com (github.com)... 52.74.223.119
Connecting to github.com (github.com)|52.74.223.119|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ErikSeguinte/movie_data/master/data/movies.pkl.xz [following]
--2020-03-06 03:41:36--  https://raw.githubusercontent.com/ErikSeguinte/movie_data/master/data/movies.pkl.xz
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7917036 (7.5M) [application/octet-stream]
Saving to: ‘movies.pkl.xz’

movies.pkl.xz       100%[===================>]   7.55M  --.-KB/s    in 0.06s   

2020-03-06 03:41:38 (131 MB/s) - ‘movies.pkl.xz’ saved [7917036/7917036]

--2020-03-06 03

## Clean Movie DF

In [4]:
movies.dtypes

adult                     object
belongs_to_collection     object
budget                    object
genres                    object
homepage                  object
id                        object
imdb_id                   object
original_language         object
original_title            object
overview                  object
popularity                object
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
dtype: object

* Movies Dataframe has malformed data. `id` Should be numeric.
* After inspection, it looks like there are rows that are missing a comma somewhere, making columns not line up, and adding the wrong data to columns. Let's clean those up.
* All malformed rows have strings for IDs instead of numeric, so we will coerce them into numeric columns, and strings will be returned as `NaN`, which we'll then drop.

* `budget` and `revanue` should also be numeric, but Nans won't be dropped






In [0]:
movies['id'] = pd.to_numeric(movies['id'], errors='coerce')
movies = movies[movies['id'].notnull()]
movies = movies.set_index('id')

In [0]:
def to_numeric(df, labels):
    
    for label in labels:
        df[label] = pd.to_numeric(movies[label], errors='coerce').copy()
    return df

In [0]:
movies = to_numeric(movies, ['budget', 'revenue', 'vote_average'])

In [0]:
movies['release_date'] =pd.to_datetime(movies['release_date'], infer_datetime_format= True)

In [0]:
movies['year'] = movies['release_date'].dt.year

In [0]:
clean_movies = movies[['title','genres', 'release_date','budget', 'revenue','year' ,'runtime', 'vote_average', 'vote_count']]

## Process User Reviews
* User reviews come in a collection of individual reviews where a review gives a movie a score of 1 to 5.
* We will take the mean ratings for each movie

In [0]:
# Aggregate mean ratings and number of votes per movie
movie_ratings =pd.DataFrame(ratings.groupby('movieId')[['rating']].agg(['mean', 'count']))['rating']
movie_ratings = movie_ratings.rename({'mean': 'rating', 'count': 'num_votes'}, axis = 1)


* Lets drop any movies with less than 100 votes. Those are more easily swayed by outliers and aren't reliable.

In [0]:
movie_ratings = movie_ratings[(movie_ratings['num_votes'] >= 100)]

* And now we merge the averaged ratings back with the movie database.
* Note that not all movies are present in the user votings.

In [0]:
clean_movies = clean_movies.merge(movie_ratings, left_index = True, right_index=True)

In [65]:
clean_movies[['title', 'rating', 'num_votes']].nlargest(10, 'rating').style.hide_index().format(
    {'rating':"{:.2f}",
     'num_votes': "{:,.0f}"}
)

* The movie Database also provides a rating and suffer from a similar problem of some movies having a tiny sample size.

In [72]:
(clean_movies[clean_movies['vote_count'] >= 100]
    [['title', 'vote_average', 'vote_count']]
    .nlargest(10, 'vote_average')
    .style
    .hide_index()
    .format({'vote_average':'{:.1f}', 'vote_count':'{:,.0f}'})
)
 


In [74]:
(clean_movies[['title', 'revenue']]
 .nlargest(10, 'revenue')
 .style
 .hide_index()
 .format({
     "revenue":"${:,.0f}"
 }
 )
)

## Inflation
* Inflation means that a 1940 dollar is worth more than a 2020 dollar. Let's adjust Revenue for that.
* The Consumer price index can be used to convert to standarized dollars.
* Here, we'll be using 2014 dollars.
* Years later than 2014 will not be adjusted.
$$ \textrm{adjusted dollars} = \frac{\textrm{New CPI}}{\textrm{Base CPI}}$$
* where x is the current cpi and y is the cpi of that year 

In [0]:
cpi = cpi[cpi['Country Name'] == 'United States'][['Year', 'CPI']]

In [0]:
cpi = cpi.set_index(cpi['Year'])

In [0]:
def adjust_dollars(value, year):
    year = int(year)
    try:
        current = cpi.loc[2014,'CPI']
        base = cpi.loc[year,'CPI']
        adjusted_value = value * (current/base)
        return adjusted_value
    except: 
        return value

In [0]:
clean_movies['year']= clean_movies['release_date'].dt.year

In [21]:
df = clean_movies[clean_movies['revenue'].notnull() & clean_movies['year'].notnull()]
df

title  ... num_votes
5.0           Four Rooms  ...     15258
6.0       Judgment Night  ...     27895
11.0           Star Wars  ...     19475
12.0        Finding Nemo  ...      4475
13.0        Forrest Gump  ...      1838
...                  ...  ...       ...
103299.0   Traces of Red  ...       141
109161.0     The Squeeze  ...       120
110669.0   Wabash Avenue  ...       173
115210.0        Stitches  ...      3144
116977.0      Foodfight!  ...       880

[1299 rows x 11 columns]

In [0]:
adjusted = pd.DataFrame([adjust_dollars(x,y) for x,y in zip(df['revenue'], df['year'])], index = df.index, columns = ['adjusted_revenue'])

In [0]:
clean_movies = clean_movies.merge(adjusted, left_index=True, right_index=True)

In [24]:
clean_movies[['title', 'adjusted_revenue']].nlargest(10, 'adjusted_revenue')

title  adjusted_revenue
11.0                                        Star Wars      3.028728e+09
597.0                                         Titanic      2.721128e+09
601.0                      E.T. the Extra-Terrestrial      1.945322e+09
1891.0                        The Empire Strikes Back      1.546830e+09
329.0                                   Jurassic Park      1.507846e+09
122.0   The Lord of the Rings: The Return of the King      1.439899e+09
238.0                                   The Godfather      1.387267e+09
1892.0                             Return of the Jedi      1.361233e+09
1893.0      Star Wars: Episode I - The Phantom Menace      1.313639e+09
671.0        Harry Potter and the Philosopher's Stone      1.305537e+09

In [0]:
clean_movies['decade'] = [x - (x%10) for x in clean_movies['year']]

In [26]:
clean_movies

title  ...  decade
5.0           Four Rooms  ...  1990.0
6.0       Judgment Night  ...  1990.0
11.0           Star Wars  ...  1970.0
12.0        Finding Nemo  ...  2000.0
13.0        Forrest Gump  ...  1990.0
...                  ...  ...     ...
103299.0   Traces of Red  ...  1990.0
109161.0     The Squeeze  ...  1980.0
110669.0   Wabash Avenue  ...  1950.0
115210.0        Stitches  ...  2010.0
116977.0      Foodfight!  ...  2010.0

[1301 rows x 13 columns]

In [27]:
#enable_plotly_in_cell()
#clean_movies.groupby('year')['vote_average'].mean().iplot(kind='bar')
clean_movies.groupby('year')['vote_average'].mean()
#px.bar(clean_movies.groupby('year')['vote_average'].mean())

year
1915.0    6.400000
1925.0    7.350000
1927.0    7.950000
1930.0    6.700000
1931.0    7.566667
            ...   
2011.0    6.393750
2012.0    5.100000
2013.0    5.400000
2014.0    6.300000
2015.0    5.700000
Name: vote_average, Length: 85, dtype: float64

In [28]:
px.bar(clean_movies.groupby('year')['vote_average'].mean().reset_index(), 
       x ='year',
       y = 'vote_average' )

In [29]:
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('default')

In [30]:
alt.Chart(clean_movies, width=1080).mark_bar().encode(
    alt.Y('mean(vote_average)'),
    alt.X('year(release_date):O')
)

alt.Chart(...)

In [31]:
alt.Chart(clean_movies, width=1080).mark_bar().encode(
    alt.Y('mean(rating)'),
    alt.X('year(release_date):N')
)

alt.Chart(...)

In [32]:
alt.Chart(clean_movies,width=720).mark_bar().encode(
    alt.Y('mean(vote_average)'),
    alt.X('decade:O')
)

alt.Chart(...)

* I'd like to compare the votes from TMB to the user ratings, but they are on different scales. We'll use standard scaler to normalize them so we can more easily compare.

In [0]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
to_scale = clean_movies[['vote_average', 'rating']].dropna()
scaled = scaler.fit_transform(to_scale)
clean_movies = clean_movies.merge(
    pd.DataFrame(
        scaled,
        index = to_scale.index,
        columns = ['scaled_tmdb_vote', 'scaled_user_rating']
    ),
    left_index = True,
    right_index = True,
)

In [34]:
traces = [
    go.Bar(name='TMDB rating',
        x = clean_movies.groupby('decade')['scaled_tmdb_vote'].mean().index,
        y = clean_movies.groupby('decade')['scaled_tmdb_vote'].mean()
    ),
        go.Bar(name='user rating',
        x = clean_movies.groupby('decade')['scaled_user_rating'].mean().index,
        y = clean_movies.groupby('decade')['scaled_user_rating'].mean()
    )
]

go.Figure(data = traces,
    layout_xaxis_tick0 = 1890
)

In [35]:
from sklearn.decomposition import PCA

pca = PCA(1)

to_scale = clean_movies[['vote_average', 'rating', 'adjusted_revenue']].dropna()
scaled = scaler.fit_transform(to_scale)

pca_df = pd.DataFrame(pca.fit_transform(scaled), index=to_scale.index, columns = ['PCA'])

clean_movies = clean_movies.merge(pca_df, left_index = True, right_index = True)
clean_movies.head(2)

title  ...       PCA
5.0      Four Rooms  ... -0.604977
6.0  Judgment Night  ... -0.533044

[2 rows x 16 columns]

In [36]:
clean_movies[['title', 'PCA']].nlargest(10, 'PCA')

title       PCA
11.0                                        Star Wars  8.865144
597.0                                         Titanic  7.507243
601.0                      E.T. the Extra-Terrestrial  5.271077
1891.0                        The Empire Strikes Back  4.819260
238.0                                   The Godfather  4.757413
122.0   The Lord of the Rings: The Return of the King  4.513990
329.0                                   Jurassic Park  4.351498
1892.0                             Return of the Jedi  4.219271
121.0           The Lord of the Rings: The Two Towers  3.981781
155.0                                 The Dark Knight  3.874508

|      | title                                         |     PCA |
|-----:|:----------------------------------------------|--------:|
|   11 | Star Wars                                     | 8.86514 |
|  597 | Titanic                                       | 7.50724 |
|  601 | E.T. the Extra-Terrestrial                    | 5.27108 |
| 1891 | The Empire Strikes Back                       | 4.81926 |
|  238 | The Godfather                                 | 4.75741 |
|  122 | The Lord of the Rings: The Return of the King | 4.51399 |
|  329 | Jurassic Park                                 | 4.3515  |
| 1892 | Return of the Jedi                            | 4.21927 |
|  121 | The Lord of the Rings: The Two Towers         | 3.98178 |
|  155 | The Dark Knight                               | 3.87451 |

In [0]:
clean_movies['q_budget'] = pd.qcut(clean_movies['budget'], q = 5, 
labels = ['vlow', 'low', 'med', 'high', 'vhigh']
)

In [0]:
budget_ratings = clean_movies[['title','q_budget', 'budget', 'revenue', 'rating', 'vote_average']].dropna()

In [39]:
px.histogram(budget_ratings.dropna(),
       x = 'q_budget',
       y = 'vote_average')

In [40]:
alt.Chart(budget_ratings, width=720).mark_bar().encode(
    x='q_budget:N',
    y='mean(vote_average)'
)

alt.Chart(...)

In [41]:
trace = go.Scatter(
    y = budget_ratings['vote_average'],
    x = budget_ratings['revenue'],
    mode = 'markers'
)

go.Figure(
    trace,
    layout_xaxis_title = "Budget",
    layout_yaxis_title = "Movie Rating",
    layout_title = "Movie Ratings by budget",
    
)

In [0]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

## Genre

In [0]:
genres = clean_movies[['title','genres']]

In [0]:
import ast

In [0]:
bad_genres = [
           'Aniplex',
 'BROSTA TV',
 'Carousel Productions',   
  'GoHands',
   'Mardock Scramble Production Committee',
    'Odyssey Media',
     'Pulser Productions',
 'Rogue State',
  'Sentai Filmworks',
   'Telescene Film Group Productions',
 'The Cartel',
  'Vision View Entertainment',
]

#genre_set = genre_set.difference(bad_genres)

In [46]:
genres['genres_ls'] = [
                    [d['name'] for d in ast.literal_eval(x) if d['name'] not in bad_genres ]
                    for x in genres['genres']
                    ]
genres['genres_ls']

5.0                                   [Crime, Comedy]
6.0                         [Action, Thriller, Crime]
11.0             [Adventure, Action, Science Fiction]
12.0                              [Animation, Family]
13.0                         [Comedy, Drama, Romance]
                              ...                    
103299.0    [Drama, Action, Thriller, Crime, Mystery]
109161.0                      [Crime, Action, Comedy]
110669.0                                      [Music]
115210.0                                     [Horror]
116977.0          [Animation, Action, Comedy, Family]
Name: genres_ls, Length: 1552, dtype: object

In [0]:
from sklearn.preprocessing import MultiLabelBinarizer

In [0]:
mlb = MultiLabelBinarizer()

In [0]:
mlb_ = mlb.fit_transform(genres['genres_ls'])

In [50]:
encoded_genres = pd.DataFrame(mlb_, columns = mlb.classes_, index = genres.index)
encoded_genres

Action  Adventure  Animation  ...  Thriller  War  Western
5.0            0          0          0  ...         0    0        0
6.0            1          0          0  ...         1    0        0
11.0           1          1          0  ...         0    0        0
12.0           0          0          1  ...         0    0        0
13.0           0          0          0  ...         0    0        0
...          ...        ...        ...  ...       ...  ...      ...
103299.0       1          0          0  ...         1    0        0
109161.0       1          0          0  ...         0    0        0
110669.0       0          0          0  ...         0    0        0
115210.0       0          0          0  ...         0    0        0
116977.0       1          0          1  ...         0    0        0

[1552 rows x 19 columns]

In [51]:
encoded_genres = clean_movies.merge(encoded_genres, **merge_keys)
encoded_genres

title  ... Western
5.0           Four Rooms  ...       0
6.0       Judgment Night  ...       0
11.0           Star Wars  ...       0
12.0        Finding Nemo  ...       0
13.0        Forrest Gump  ...       0
...                  ...  ...     ...
103299.0   Traces of Red  ...       0
109161.0     The Squeeze  ...       0
110669.0   Wabash Avenue  ...       0
115210.0        Stitches  ...       0
116977.0      Foodfight!  ...       0

[66832 rows x 36 columns]

In [52]:
for genre in mlb.classes_:
    print(genre, encoded_genres[encoded_genres[genre] == 1]['vote_average'].mean())

Action 6.368047337278103
Adventure 6.464166666666668
Animation 6.5181818181818185
Comedy 6.599013956303933
Crime 6.600525787915338
Documentary 7.06
Drama 6.6025728133283295
Family 6.493814432989691
Fantasy 6.527692307692308
Foreign 6.0
History 6.9728813559322065
Horror 6.329523809523811
Music 6.574999999999999
Mystery 6.828985507246377
Romance 6.59998176041643
Science Fiction 6.434319526627218
Thriller 6.599510324119252
War 7.116279069767442
Western 7.095238095238094


In [0]:
d = {g: [encoded_genres[encoded_genres[g] == 1]['PCA'].mean(),
         encoded_genres[encoded_genres[g] == 1]['adjusted_revenue'].sum()]
     for g in mlb.classes_}

In [54]:
df = pd.DataFrame.from_dict(d, orient='index',  columns = ['pca rating', 'total revenue'])
df

pca rating  total revenue
Action             0.020301   9.119107e+10
Adventure          0.378806   9.042168e+10
Animation          0.261648   1.372190e+10
Comedy            -0.347387   2.908821e+12
Crime             -0.346864   2.883037e+12
Documentary       -0.085474   2.820166e+08
Drama             -0.344216   2.932161e+12
Family             0.328039   3.411401e+10
Fantasy            0.308037   4.323811e+10
Foreign           -1.046721   4.106328e+06
History            0.190439   8.390535e+09
Horror            -0.466021   1.060188e+10
Music             -0.163118   6.592343e+09
Mystery            0.025770   1.785842e+10
Romance           -0.347252   2.883544e+12
Science Fiction    0.074629   4.584190e+10
Thriller          -0.346721   2.917061e+12
War                0.491545   8.984297e+09
Western            0.205011   2.296981e+09

In [55]:
trace = go.Bar(
    x=df.index,
    y=df['pca rating']
)

fig = go.Figure(
    data = trace,
    #layout_x)
)
fig.update_xaxes(tickangle=45)
fig.update_layout(
    title_text = "PCA Rating by Genre"
)

In [56]:
melted = df.reset_index().melt(id_vars='index')
melted

index       variable         value
0            Action     pca rating  2.030121e-02
1         Adventure     pca rating  3.788064e-01
2         Animation     pca rating  2.616476e-01
3            Comedy     pca rating -3.473871e-01
4             Crime     pca rating -3.468643e-01
5       Documentary     pca rating -8.547420e-02
6             Drama     pca rating -3.442159e-01
7            Family     pca rating  3.280390e-01
8           Fantasy     pca rating  3.080365e-01
9           Foreign     pca rating -1.046721e+00
10          History     pca rating  1.904386e-01
11           Horror     pca rating -4.660211e-01
12            Music     pca rating -1.631184e-01
13          Mystery     pca rating  2.576993e-02
14          Romance     pca rating -3.472515e-01
15  Science Fiction     pca rating  7.462878e-02
16         Thriller     pca rating -3.467213e-01
17              War     pca rating  4.915449e-01
18          Western     pca rating  2.050108e-01
19           Action  total revenue  9.119107e+10
20        Adventure  total revenue  9.042168e+10
21        Animation  total revenue  1.372190e+10
22           Comedy  total revenue  2.908821e+12
23            Crime  total revenue  2.883037e+12
24      Documentary  total revenue  2.820166e+08
25            Drama  total revenue  2.932161e+12
26           Family  total revenue  3.411401e+10
27          Fantasy  total revenue  4.323811e+10
28          Foreign  total revenue  4.106328e+06
29          History  total revenue  8.390535e+09
30           Horror  total revenue  1.060188e+10
31            Music  total revenue  6.592343e+09
32          Mystery  total revenue  1.785842e+10
33          Romance  total revenue  2.883544e+12
34  Science Fiction  total revenue  4.584190e+10
35         Thriller  total revenue  2.917061e+12
36              War  total revenue  8.984297e+09
37          Western  total revenue  2.296981e+09